internal/pkg/poster/poster.go

```diff
@@ -0,0 +1,261 @@
+)
+
+// TODO: replace with state.NewStore
+type actorStore struct {
+	ctx context.Context
+	cst.ChainStateReadWriter
+}
+
+func (as *actorStore) Context() context.Context {
+	return as.ctx
+}
+
+var _ adt.Store = new(actorStore)
+
+// Poster listens for changes to the chain head and generates and submits a PoSt if one is required.
+type Poster struct {
+	postMutex      sync.Mutex
+	postCancel     context.CancelFunc
+	scheduleCancel context.CancelFunc
+
+	minerAddr        address.Address
+	outbox           *message.Outbox
+	sectorbuilder    sectorbuilder.Interface
+	minerNode        *storageminerconnector.StorageMinerNodeConnector
+	storageMiner     *storage.Miner
+	heaviestTipsetCh chan interface{}
+	chain            *cst.ChainStateReadWriter
+	stateViewer      *appstate.Viewer
+}
+
+// NewPoster creates a Poster struct
+func NewPoster(
+	minerAddr address.Address,
+	outbox *message.Outbox,
+	sb sectorbuilder.Interface,
+	connector *storageminerconnector.StorageMinerNodeConnector,
+	miner *storage.Miner,
+	htc chan interface{},
+	chain *cst.ChainStateReadWriter,
+	stateViewer *appstate.Viewer) *Poster {
+
+	return &Poster{
+		minerAddr:        minerAddr,
+		outbox:           outbox,
+		sectorbuilder:    sb,
+		minerNode:        connector,
+		storageMiner:     miner,
+		heaviestTipsetCh: htc,
+		chain:            chain,
+		stateViewer:      stateViewer,
+	}
+}
+
+// StartPoSting starts the posting scheduler
+func (p *Poster) StartPoSting(ctx context.Context) {
+	p.StopPoSting()
+	ctx, p.scheduleCancel = context.WithCancel(ctx)
+
+	go func() {
+		for {
+			select {
+			case ts, ok := <-p.heaviestTipsetCh:
+				if !ok {
+					return
+				}
+
+				newHead, ok := ts.(block.TipSet)
+				if !ok {
+					log.Warn("non-tipset published on heaviest tipset channel")
+					continue
+				}
+
+				err := p.startPoStIfNeeded(ctx, newHead)
+				if err != nil {
+					log.Error("error attempting fallback post", err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// StopPoSting stops the posting scheduler if running and any outstanding PoSts.
+func (p *Poster) StopPoSting() {
+	p.postMutex.Lock()
+	defer p.postMutex.Unlock()
+
+	if p.scheduleCancel != nil {
+		p.postCancel()
+
+		p.scheduleCancel()
+		p.scheduleCancel = nil
+	}
+}
+
+func (p *Poster) startPoStIfNeeded(ctx context.Context, newHead block.TipSet) error {
+	p.postMutex.Lock()
+	defer p.postMutex.Unlock()
+
+	if p.postCancel != nil {
+		// already posting
+		return nil
+	}
+
+	tipsetHeight, err := newHead.Height()
+	if err != nil {
+		return err
+	}
+
+	root, err := p.chain.GetTipStateStateRoot(ctx, newHead.Key())
+	if err != nil {
+		return err
+	}
+
+	stateView := p.stateViewer.StateView(root)
+	provingPeriodStart, _, _, err := stateView.MinerProvingPeriod(ctx, p.minerAddr)
+	if err != nil {
+		return nil
+	}
+
+	if uint64(provingPeriodStart) <= tipsetHeight {
+		// it's not time to PoSt
+		return nil
+	}
+
+	ctx, p.postCancel = context.WithCancel(ctx)
+	go func() {
```

Comment:
nit: stylistic preference to define this as a method on Poster i.e. `doPoSt(ctx, ppStart, stateView)` and then call it in a go routine here.  Suggesting only to keep function sizes down.